File: engine/Arduino/vgm_game_engine/src/internal/level.cpp
#include "level.h"
#include "game.h"
#include "entity.h"

namespace VGMGameEngine
{
    // Default Constructor
    Level::Level()
        : name(""),
          size(Vector(0, 0)),
          game(nullptr),
          _start(nullptr),
          _stop(nullptr),
          entity_count(0),
          entities(nullptr)
    {
    }

    // Parameterized Constructor
    Level::Level(const char *name, const Vector &size, Game *game, void (*start)(Level &), void (*stop)(Level &))
        : name(name),
          size(size),
          game(game),
          _start(start),
          _stop(stop),
          entity_count(0),
          entities(nullptr)
    {
    }

    // Destructor
    Level::~Level()
    {
        clear();
    }

    // Clear all entities
    void Level::clear()
    {
        for (int i = 0; i < entity_count; i++)
        {
            if (entities[i] != nullptr)
            {
                entities[i]->stop(this->game);
                delete entities[i];
                entities[i] = nullptr;
            }
        }
        // Free the dynamic array
        delete[] entities;
        entities = nullptr;
        entity_count = 0;
    }

    // Get list of collisions for a given entity
    Entity **Level::collision_list(Entity *entity, int &count) const
    {
        count = 0;
        if (entity_count == 0)
        {
            return nullptr;
        }

        Entity **result = new Entity *[entity_count];
        for (int i = 0; i < entity_count; i++)
        {
            if (entities[i] != nullptr &&
                entities[i] != entity && // Skip self
                is_collision(entity, entities[i]))
            {
                result[count++] = entities[i];
            }
        }
        return result;
    }

    // Add an entity to the level
    void Level::entity_add(Entity *entity)
    {
        // Allocate a new array with size one greater than the current count
        Entity **newEntities = new Entity *[entity_count + 1];
        // Copy the existing entity pointers (if any)
        for (int i = 0; i < entity_count; i++)
        {
            newEntities[i] = entities[i];
        }
        newEntities[entity_count] = entity;

        // Delete the old array
        delete[] entities;
        entities = newEntities;
        entity_count++;

        // Start the new entity
        entity->start(this->game);
        entity->is_active = true;
    }

    // Remove an entity from the level
    void Level::entity_remove(Entity *entity)
    {
        if (entity_count == 0)
            return;

        int remove_index = -1;
        for (int i = 0; i < entity_count; i++)
        {
            if (entities[i] == entity)
            {
                remove_index = i;
                break;
            }
        }
        if (remove_index == -1)
            return;

        // Stop and delete the entity
        entities[remove_index]->stop(this->game);
        delete entities[remove_index];

        // Allocate a new array with one fewer slot (if any remain)
        Entity **newEntities = (entity_count - 1 > 0) ? new Entity *[entity_count - 1] : nullptr;
        // Copy over all pointers except the removed one
        for (int i = 0, j = 0; i < entity_count; i++)
        {
            if (i == remove_index)
                continue;
            newEntities[j++] = entities[i];
        }

        // Free the old array and update state
        delete[] entities;
        entities = newEntities;
        entity_count--;
    }

    // Check if any entity has collided with the given entity
    bool Level::has_collided(Entity *entity) const
    {
        for (int i = 0; i < entity_count; i++)
        {
            if (entities[i] != nullptr &&
                entities[i] != entity &&
                is_collision(entity, entities[i]))
            {
                return true;
            }
        }
        return false;
    }

    // Determine if two entities are colliding
    bool Level::is_collision(const Entity *a, const Entity *b) const
    {
        return a->position.x < b->position.x + b->size.x &&
               a->position.x + a->size.x > b->position.x &&
               a->position.y < b->position.y + b->size.y &&
               a->position.y + a->size.y > b->position.y;
    }

    // Render all active entities
    void Level::render(Game *game)
    {
        if (game->is_8bit)
        {
            // clear screen
            game->draw->clear(Vector(0, 0), game->size, game->bg_color);
        }

        for (int i = 0; i < entity_count; i++)
        {
            Entity *ent = entities[i];
            if (ent != nullptr && ent->is_active)
            {
                if (!game->is_8bit)
                {
                    float old_screen_x = ent->old_position.x - game->old_pos.x;
                    float old_screen_y = ent->old_position.y - game->old_pos.y;

                    if (!(old_screen_x + ent->size.x < 0 || old_screen_x > game->size.x ||
                          old_screen_y + ent->size.y < 0 || old_screen_y > game->size.y))
                    {
                        game->draw->clear(Vector(old_screen_x, old_screen_y), Vector(ent->size.x, ent->size.y), game->bg_color);
                    }
                }

                ent->render(game->draw, game);

                if (!game->is_8bit)
                {
                    if (ent->sprite != nullptr)
                    {
                        game->draw->image(Vector(ent->position.x - game->pos.x, ent->position.y - game->pos.y), ent->sprite);
                    }
                }
                else
                {
                    // draw 8bit sprite
                    game->draw->image(Vector(ent->position.x - game->pos.x, ent->position.y - game->pos.y), ent->sprite->data, ent->sprite->size);
                }
            }
        }

        if (game->is_8bit)
        {
            // send newly drawn pixels to the display
            game->draw->swap();
        }
    }

    // Start the level
    void Level::start()
    {
        if (_start != nullptr)
        {
            _start(*this);
        }
    }

    // Stop the level
    void Level::stop()
    {
        if (_stop != nullptr)
        {
            _stop(*this);
        }
    }

    // Update all active entities
    void Level::update(Game *game)
    {
        for (int i = 0; i < entity_count; i++)
        {
            Entity *ent = entities[i];
            if (ent != nullptr && ent->is_active)
            {
                ent->update(this->game);
                int count = 0;
                Entity **collisions = collision_list(ent, count);
                for (int j = 0; j < count; j++)
                {
                    ent->collision(collisions[j], this->game);
                }
                delete[] collisions;
            }
        }
    }
}

File: src/CircuitPython/src/pico_game_engine/level.py
from picogui.vector import Vector
from .entity import Entity


class Level:
    """
    Represents a level in the game.

    Parameters:
    - name: str - the name of the level
    - size: Vector - the size of the level
    - game: Game - the game to which the level belongs
    - start: function(Level) - the function called when the level is created
    - stop: function(Level) - the function called when the level is destroyed
    """

    def __init__(
        self,
        name: str,
        size: Vector,
        game,
        start=None,  # start is a function that is called when the level is created
        stop=None,  # stop is a function that is called when the level is destroyed
    ):
        self.name = name
        self.size = size
        self.game = game
        self.entities: list[Entity] = []  # List of entities in the level
        self._start = start
        self._stop = stop

    def clear(self):
        """Clear the level"""
        self.entities = []

    def collision_list(self, entity: Entity) -> list:
        """Return a list of entities that the entity collided with"""
        collided = []
        for other in self.entities:
            if entity != other and self.is_collision(entity, other):
                collided.append(other)
        return collided

    def entity_add(self, entity: Entity):
        """Add an entity to the level"""
        self.entities.append(entity)
        entity.start(self.game)
        entity.is_active = True

    def entity_remove(self, entity: Entity):
        """Remove an entity from the level"""
        self.entities.remove(entity)

    def has_collided(self, entity: Entity) -> bool:
        """Check for collisions with other entities"""
        for other in self.entities:
            if entity != other and self.is_collision(entity, other):
                return True
        return False

    def is_collision(self, entity: Entity, other: Entity) -> bool:
        """Check if two entities collided using AABB logic"""
        entity_pos = entity.position
        other_pos = other.position
        entity_size = entity.size
        other_size = other.size
        return (
            entity_pos.x < other_pos.x + other_size.x
            and entity_pos.x + entity_size.x > other_pos.x
            and entity_pos.y < other_pos.y + other_size.y
            and entity_pos.y + entity_size.y > other_pos.y
        )

    def render(self):
        """Render the level"""
        for entity in self.entities:
            if entity.is_active:
                entity.render(self.game.draw, self.game)
                if entity.tile_grid and entity.sprite_path != "":
                    self.game.draw.tile_grid(
                        Vector(
                            entity.position.x - self.game.position.x,
                            entity.position.y - self.game.position.y,
                        ),
                        entity.tile_grid,
                    )
        self.game.draw.swap()

    def start(self):
        """Start the level"""
        if self._start:
            self._start(self)

    def stop(self):
        """Stop the level"""
        if self._stop:
            self._stop(self)

    def update(self):
        """Update the level"""
        for entity in self.entities:
            if entity.is_active:
                entity.update(self.game)
                collided = self.collision_list(entity)
                for other in collided:
                    entity.collision(other, self.game)
